# **Finding the zeros of a quadratic function:**

In this problem, I am solving a simple regression problem, where points are coming from the line  𝑦=2𝑥+1.

I will: <br>

> (i) Generate data points coming from a quadratic function:  𝑦=−𝑥2+3𝑥+10 , plus some noise. <br>

> (ii) Modify a PyTorch model in order for it to learn a quadratic function of the form  𝑦=𝑎𝑥2+𝑏𝑥+𝑐 . <br>


> (iii) Train my model and report what values it computes.



In [1]:
import torch.optim as optim
import torch.nn as nn
import numpy as np
import torch

In [2]:
#(i) Generating data points coming from a quadratic function:  𝑦=−𝑥**2+3𝑥+10 , plus some noise.

np.random.seed(42)
x = np.random.rand(100, 1)
y = -x**2 + 3 * x + 10 + .1 * np.random.randn(100, 1)

In [3]:
# Shuffling the indices
idx = np.arange(100)
np.random.shuffle(idx)

# Using the first 80 random indices for train
train_idx = idx[:80]

# Using the remaining indices for validation
val_idx = idx[80:]

# Generating train and validation sets
x_train, y_train = x[train_idx], y[train_idx]
x_val, y_val = x[val_idx], y[val_idx]

In [4]:
device = 'cuda' if torch.cuda.is_available() else 'cpu'

# the data points needs to be converted from numpy array to Tensor objects

x_train_tensor = torch.from_numpy(x_train).float().to(device)
y_train_tensor = torch.from_numpy(y_train).float().to(device)

In [5]:
#(ii) Modifying the PyTorch model in order for it to learn a quadratic function of the form  𝑦=𝑎𝑥**2+𝑏𝑥+𝑐 .

class QuadraticRegression(nn.Module):
    def __init__(self):
        super().__init__()
        # To make "a","b" and "c" real parameters of the model, wrapping them with nn.Parameter
        self.a = nn.Parameter(torch.randn(1, requires_grad=True, dtype=torch.float))
        self.b = nn.Parameter(torch.randn(1, requires_grad=True, dtype=torch.float))
        self.c = nn.Parameter(torch.randn(1, requires_grad=True, dtype=torch.float))

    def forward(self, x):
        # Computing the outputs / predictions based on the quadratic functions
        return self.a * x**2 + self.b * x + self.c

In [6]:
#(iii) Training the model and reporting what values it computes.

torch.manual_seed(42)

# Creating the model and sending it to the device
model = QuadraticRegression().to(device)

#defining the learning rate
lr = 1e-1
n_epochs = 20000

#defining loss function and optimizer
loss_fn = nn.MSELoss(reduction='mean')
optimizer = optim.SGD(model.parameters(), lr=lr)

for epoch in range(n_epochs):
    # This sets the model in 'training' mode.
    model.train()
    yhat = model(x_train_tensor)

    loss = loss_fn(y_train_tensor, yhat)
    loss.backward()
    optimizer.step()
    optimizer.zero_grad()

In [7]:
#printing the result

print(model.state_dict())

OrderedDict([('a', tensor([-0.8111])), ('b', tensor([2.7849])), ('c', tensor([10.0506]))])


# **Answer:** <br>

As we can see, our a = -0.811 , b = 2.7849, and c = 10.0506.

<br>

With rounding, these values equal to -1, 3, 10 and are the real zeroes of our quadratic function.

<br>
Graphically, the real zeros of a function are the x-intercepts of a function, where the graph of the function crosses the x-axis.